jupyter/notebooks/Data Traversal.ipynb

```diff
@@ -0,0 +1,583 @@
+   "id": "81081ece-a46c-44fc-ac4a-241afa20d22b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "first_result = response['hits']['hits'][0]\n",
+    "print(first_result['_id'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a15739d6-cb37-403e-aeb6-76138a9d2b77",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#checking the image\n",
+    "iiif_url = first_result['_source']['state']['derivedData']['thumbnail']['url']\n",
+    "image_url = str(IIIFImageClient().init_from_url(iiif_url).size(width=500))\n",
+    "Image.open(BytesIO(httpx.get(image_url).content))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "17715f49-6784-48ee-ac68-ec645d6356a7",
+   "metadata": {},
+   "source": [
+    "Check if the image sorted as expected."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7fe57b4e-96d9-4013-a7dd-f47a9b8b6ea9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "response = local_es.search(\n",
+    "    index=os.environ['INDEX_NAME'],\n",
+    "    body={\n",
+    "        \"query\":{\n",
+    "            \"match_all\":{}\n",
+    "        },\n",
+    "        \"size\":30,\n",
+    "        \"from\":0,\n",
+    "        \"search_after\": ['a24yqdyw'],\n",
+    "        \"sort\":{\n",
+    "            \"_id\":\"desc\", #descending order\n",
+    "        },\n",
+    "    }\n",
+    ")\n",
+    "\n",
+    "first_result = response['hits']['hits'][0]\n",
+    "print(first_result['_id'])\n",
+    "#checking the image\n",
+    "iiif_url = first_result['_source']['state']['derivedData']['thumbnail']['url']\n",
+    "image_url = str(IIIFImageClient().init_from_url(iiif_url).size(width=500))\n",
+    "Image.open(BytesIO(httpx.get(image_url).content))\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e93d1400-0e77-46a1-87cb-2af53a104a2e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\n",
+    "    f\"Found {response['hits']['total']['value']} \"\n",
+    "    f\"results in {response['took'] / 1000}s\"\n",
+    ")\n",
+    "\n",
+    "print()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7efb25b2-f61f-44bc-b7ef-74c10b9d4ff8",
+   "metadata": {},
+   "source": [
+    "This is weird. Why its still getting 10k results? There should only be 20."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b996359-c8b3-49c2-aac4-1d13e63d337f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(len(response['hits']['hits']))\n",
+    "last_result = response['hits']['hits'][20]\n",
+    "print(last_result['_id'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6c4c9bc7-3ac6-4f70-95cc-f410d7eebd78",
+   "metadata": {},
+   "source": [
+    "The hits does only have 20 elements. I'm confused. Is there any tool that I can use check the response JSON in a human eye frinedly way? (not the leeeeengthy data XD)"
```

Comment:
Ah yeah, this can be a bit confusing...

`hits` describes the number of documents in the index which matched your query, with a default upper limit of 10,000. By setting the `size`, you're just limiting the number of documents which will actually be _returned_ for your request, but the `hits` will still describe the size of everything which matched...

The upper limit on matched documents is a setting which you can change, or you can use `search_after` to fetch documents in batches (search_after doesn't care about the 10,000 limit). 

[From the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html):

> By default, you cannot use from and size to page through more than 10,000 hits. This limit is a safeguard set by the index.max_result_window index setting. If you need to page through more than 10,000 hits, use the search_after parameter instead.

---

There's not much you can do to make the json response more friendly to look at, sadly. 

Some options:
- If you know the already fields you want to see in the response, you can add `_source=["a_field_which_you_want_to_see_in_the_response", "and_another_one"]` to your search to cut out the most annoying, useless fields - that alone should make the response more manageable. 
- Sometimes it's useful to copy the json out of jupyter and into a "proper" text editor (like vscode or pycharm) so that you can edit it, run complex searches etc
- Sometimes it's more useful to look at the raw data in kibana...

I still don't have a perfect solution for this myself tbh!